File: Sources/AgentManagerCore/Agent.swift
import Foundation
import Yams

public enum TriggerType: String, Codable {
    case schedule
    case manual
    case fileWatch = "file-watch"
}

public struct Trigger: Codable {
    public let type: TriggerType
    public let hour: Int?
    public let minute: Int?
    public let watchPath: String?

    public init(type: TriggerType, hour: Int? = nil, minute: Int? = nil, watchPath: String? = nil) {
        self.type = type
        self.hour = hour
        self.minute = minute
        self.watchPath = watchPath
    }

    enum CodingKeys: String, CodingKey {
        case type
        case hour
        case minute
        case watchPath = "watch_path"
    }
}

public struct Agent: Codable {
    public let name: String
    public let description: String
    public let trigger: Trigger
    public let workingDirectory: String
    public let contextScript: String?
    public let prompt: String
    public let allowedTools: [String]
    public let maxTurns: Int
    public let maxBudgetUSD: Double

    public init(
        name: String,
        description: String,
        trigger: Trigger,
        workingDirectory: String,
        contextScript: String? = nil,
        prompt: String,
        allowedTools: [String],
        maxTurns: Int = 10,
        maxBudgetUSD: Double = 1.0
    ) {
        self.name = name
        self.description = description
        self.trigger = trigger
        self.workingDirectory = workingDirectory
        self.contextScript = contextScript
        self.prompt = prompt
        self.allowedTools = allowedTools
        self.maxTurns = maxTurns
        self.maxBudgetUSD = maxBudgetUSD
    }

    enum CodingKeys: String, CodingKey {
        case name
        case description
        case trigger
        case workingDirectory = "working_directory"
        case contextScript = "context_script"
        case prompt
        case allowedTools = "allowed_tools"
        case maxTurns = "max_turns"
        case maxBudgetUSD = "max_budget_usd"
    }

    /// Expand ~ in working directory to full path
    public var expandedWorkingDirectory: String {
        if workingDirectory.hasPrefix("~/") {
            return FileManager.default.homeDirectoryForCurrentUser.path + String(workingDirectory.dropFirst())
        }
        return workingDirectory
    }
}

// MARK: - YAML Parsing

public extension Agent {
    static func load(from data: Data) throws -> Agent {
        let decoder = YAMLDecoder()
        return try decoder.decode(Agent.self, from: data)
    }

    static func load(fromFile path: String) throws -> Agent {
        let url = URL(fileURLWithPath: path)
        let data = try Data(contentsOf: url)
        return try load(from: data)
    }

    func toYAML() throws -> String {
        let encoder = YAMLEncoder()
        return try encoder.encode(self)
    }

    func save(toFile path: String) throws {
        let yaml = try toYAML()
        let url = URL(fileURLWithPath: path)
        try yaml.write(to: url, atomically: true, encoding: .utf8)
    }
}

// MARK: - Template Generation

public extension Agent {
    static func template(name: String) -> Agent {
        Agent(
            name: name,
            description: "Description of what this agent does",
            trigger: Trigger(type: .manual),
            workingDirectory: "~/",
            contextScript: """
                # Gather context for the agent
                echo "Current directory: $(pwd)"
                echo "Date: $(date)"
                """,
            prompt: """
                You are an automated agent. Your task is to...

                Please analyze the context provided and take appropriate action.
                """,
            allowedTools: ["Read", "Edit", "Write", "Bash(git *)"],
            maxTurns: 10,
            maxBudgetUSD: 1.0
        )
    }
}

File: python/agent_manager/models/agent.py
"""Agent model - reusable prompt templates."""

from __future__ import annotations

from enum import Enum
from pathlib import Path
import yaml
from typing_extensions import Self
from pydantic import BaseModel, Field


class TriggerType(str, Enum):
    """Type of trigger for an agent."""

    SCHEDULE = "schedule"
    MANUAL = "manual"
    FILE_WATCH = "file-watch"


class Trigger(BaseModel):
    """Trigger configuration for an agent."""

    type: TriggerType
    hour: int | None = None
    minute: int | None = None
    watch_path: str | None = None


class Agent(BaseModel):
    """A reusable agent template with default settings."""

    name: str
    description: str
    trigger: Trigger | None = None  # Optional - agents are templates used in workflows
    working_directory: str = Field(default="~/", alias="working_directory")
    context_script: str | None = Field(default=None, alias="context_script")
    prompt: str
    allowed_tools: list[str] = Field(alias="allowed_tools")
    max_turns: int = Field(default=10, alias="max_turns")
    max_budget_usd: float = Field(default=1.0, alias="max_budget_usd")

    model_config = {
        "populate_by_name": True,
        "extra": "forbid",
    }

    @property
    def expanded_working_directory(self) -> Path:
        """Expand ~ in working directory to full path."""
        return Path(self.working_directory).expanduser()

    @classmethod
    def load(cls, path: Path | str) -> Self:
        """Load an agent from a YAML file."""
        path = Path(path)
        with path.open() as f:
            data = yaml.safe_load(f)
        return cls.model_validate(data)

    @classmethod
    def from_yaml(cls, content: str) -> Self:
        """Parse an agent from YAML string."""
        data = yaml.safe_load(content)
        return cls.model_validate(data)

    def to_yaml(self) -> str:
        """Serialize agent to YAML string."""
        data = self.model_dump(by_alias=True, exclude_none=True)
        return yaml.dump(data, default_flow_style=False, sort_keys=False, allow_unicode=True)

    def save(self, path: Path | str) -> None:
        """Save agent to a YAML file."""
        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(self.to_yaml())

    @classmethod
    def template(cls, name: str) -> Self:
        """Create a template agent with default values."""
        return cls(
            name=name,
            description="Description of what this agent does",
            trigger=Trigger(type=TriggerType.MANUAL),
            working_directory="~/",
            context_script="""\
# Gather context for the agent
echo "Current directory: $(pwd)"
echo "Date: $(date)"
""",
            prompt="""\
You are an automated agent. Your task is to...

Please analyze the context provided and take appropriate action.
""",
            allowed_tools=["Read", "Edit", "Write", "Bash(git *)"],
            max_turns=10,
            max_budget_usd=1.0,
        )
